### PR TITLE
TST: _lib: improve array API assertions

### PR DESCRIPTION
**Pull request by lucascolley**
#### Reference issue
https://github.com/scipy/scipy/pull/18930#issuecomment-1705759257 @mdhaber 

#### What does this implement/fix?
- the assertions are renamed with the `xp_` prefix to avoid conflict with `np.testing`.
- `assert actual.dtype == desired.dtype` is added. This is already checked by default for PyTorch in `assert_close` and `assert_equal`.

### CODE REVIEW

**Comment by mdhaber** [comment]
LGTM but let's see what others think of the name and whether we can add a shape check.

**Comment by lucascolley** [comment]
> whether we can add a shape check

Ah forgot to mention, I think that this already happens. I checked through all the docs, e.g. https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html where it is explicitly stated. I think that it's implied for `torch` too https://pytorch.org/docs/stable/testing.html?

**Comment by mdhaber** [comment]
It doesn't for NumPy when the right hand side is a scalar.

**Comment by tylerjereddy** [comment]
https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_array_equal.html has a `strict` keyword that gets part of the way to what Matt wants, I think? I still like the "vectorized by default," but I suppose we could do something like that here if we really need it?

File: scipy/_lib/_array_api.py [hunk 1]
```diff
@@ -177,42 +177,48 @@ def is_torch(xp):
...
     if xp is None:
         xp = array_namespace(actual)
     if is_cupy(xp):
+        assert actual.dtype == desired.dtype
         return xp.testing.assert_array_equal(actual, desired, err_msg=err_msg)
     elif is_torch(xp):
         # PyTorch recommends using `rtol=0, atol=0` like this
         # to test for exact equality
         return xp.testing.assert_close(actual, desired, rtol=0, atol=0,
                                        msg=err_msg)
```
Comment (mdhaber) [inline scipy/_lib/_array_api.py:190]:
```suggestion
                                       msg=err_msg, equal_nan=True)
```
I think this will fix https://github.com/scipy/scipy/pull/18930#discussion_r1315898607

File: scipy/_lib/_array_api.py [hunk 2]
```diff
@@ -177,42 +177,48 @@ def is_torch(xp):
...
+def xp_assert_close(actual, desired, rtol=1e-07, atol=0, err_msg='', xp=None):
     if xp is None:
         xp = array_namespace(actual)
     if is_cupy(xp):
+        assert actual.dtype == desired.dtype
         return xp.testing.assert_allclose(actual, desired, rtol=rtol,
                                           atol=atol, err_msg=err_msg)
     elif is_torch(xp):
         return xp.testing.assert_close(actual, desired, rtol=rtol,
                                        atol=atol, msg=err_msg)
```
Comment (mdhaber) [inline scipy/_lib/_array_api.py:204]:
```suggestion
                                       atol=atol, msg=err_msg, equal_nan=True)
```

**Comment by lucascolley** [comment]
In addition to the naming and shape check, we have the option of checking whether the correct namespace is returned inside these functions to consider, as discussed in gh-18930.

**Comment by rgommers** [comment]
+1 for the stricter checks, and +1 for the renames.

**Comment by lucascolley** [comment]
@mdhaber does Tyler's comment https://github.com/scipy/scipy/pull/19186#issuecomment-1706965091 help address your strictness concerns?

Also happy to go with your decision on whether or not to include the namespace checks in this PR or not.

**Comment by mdhaber** [comment]
@lucascolley The comment is certainly related - `assert_array_equal` with `strict=True` would do what I think our `xp_assert_equal` should do by default (minus the additional array type check, which I wouldn't expect from a function developed before the Array API). The fact that `assert_array_equal` has this option doesn't resolve the concern, though. Rather, it gives support to the idea that these stricter checks should be added. Actually, see the discussion in numpy/numpy#21595, where these checks were added to NumPy.

https://github.com/scipy/scipy/pull/19186#issuecomment-1710463946 indicated +1 on the strictness checks, so I'd suggest for all three assertions:
- add the shape check
- add the array type check (this should also fail if the reference is a NumPy scalar but the result is a 0d array or vice-versa)
- ensure that the `AssertionError` message identifies which check failed
- (optional) add parameters to turn them off individually (e.g. `check_dtype=True`, `check_shape=True`, `check_xp=True`), or these can be added later if the need arises

As for why these should be strict by default: these checks are not performed strictly by default now, yet the manual additional of the strict checks is often neglected, and there are several inconsistencies/shortcomings/bugs in SciPy that might have been caught had the tests been stricter. So even if these stricter checks are *only* performed where these `xp_assert`s are used, I think we would be better off than we are now. (And actually, when developers see these stricter assertions fail, they might be reminded of the other places the assertions are needed.)

**Comment by rgommers** [comment]
> (And actually, when developers see these stricter assertions fail, they might be reminded of the other places the strict checks are needed.)

To confirm that: this was my exact experience with the one test on Lucas' `fft` branch that I tried to refactor.

**Comment by lucascolley** [comment]
Sorry, the rebase is messy since I failed to `git submodule update` - fixing now.

**Comment by lucascolley** [comment]
@mdhaber how does this look? Here are the `AssertionError` messages:

```python
In [1]: from scipy._lib._array_api import xp_assert_equal

In [2]: import numpy as np

In [3]: import torch

In [4]: xp_assert_equal(np.asarray(0), np.asarray([0, 0]))
AssertionError: Array shapes are not equal: actual shape = (), desired shape = (2,)

In [5]: xp_assert_equal(np.asarray([0, 0]), np.asarray(0))
AssertionError: Array shapes are not equal: actual shape = (2,), desired shape = ()

In [6]: xp_assert_equal(np.asarray(0), np.asarray(0)[()])
AssertionError: Desired a numpy scalar, but 0-D array given.

In [7]: xp_assert_equal(np.asarray(0)[()], np.asarray(0))
AssertionError: Desired a 0-D array, but numpy scalar given.

In [8]: xp_assert_equal(0, np.asarray(0)[()])
ValueError: Inputs should be arrays with a dtype attribute, python scalars and arrays are not accepted.

In [9]: xp_assert_equal(np.asarray(0)[()], 0)
ValueError: Inputs should be arrays with a dtype attribute, python scalars and arrays are not accepted.

In [10]: xp_assert_equal(np.asarray([1, 2]), torch.asarray([1, 2]))
AssertionError: Namespaces do not match: scipy._lib.array_api_compat.array_api_compat.numpy and scipy._lib.array_api_compat.array_api_compat.torch

In [11]: xp_assert_equal(np.asarray([1, 2], dtype=np.float32), np.asarray([1, 2], dtype=np.float64))
AssertionError: Desired dtype: float64, actual dtype: float32

In [12]: xp_assert_equal(np.asarray([1, 2], dtype=np.float64), np.asarray([1, 2], dtype=np.float32))
AssertionError: Desired dtype: float32, actual dtype: float64

In [13]: xp_assert_equal(np.asarray([1, 2]), np.asarray([1, 2]))
# passes
```

File: scipy/_lib/_array_api.py [hunk 3]
```diff
@@ -191,33 +192,81 @@ def is_torch(xp):
...
+                f"Desired dtype: {desired.dtype}, actual dtype: {actual.dtype}")
+    if check_shape:
+        msg = (f"Array shapes are not equal: actual shape = {actual.shape}, "
+               f"desired shape = {desired.shape}")
+        assert_(actual.shape == desired.shape, msg)
+
+
+def _assert_matching_namespace(actual, desired):
+    desired_space = array_namespace(desired)
+    if isinstance(actual, tuple):
```
Comment (mdhaber) [inline scipy/_lib/_array_api.py:215]: Is `_assert_matching_namespace` going to be used elswhere with tuple `actual`? The way it's used here, it looks like it will always have a `dtype` attribute.

If so, I'd suggest `actual = (actual,) if not isinstance(actual, tuple) else actual` or similar. Then you don't need to copy-paste the logic. Besides the simplification, it would avoid inconsistencies - the assertions are written differently, and one has a message while the other doesn't. (Or was this intentional?)

File: scipy/_lib/_array_api.py [hunk 4]
```diff
@@ -191,33 +192,81 @@ def is_torch(xp):
...
+        for arr in actual:
+            arr_space = array_namespace(arr)
+            assert arr_space == desired_space
+            if is_numpy(arr_space):
+                _check_scalar(actual, desired)
+    else:
+        actual_space = array_namespace(actual)
+        assert_(actual_space == desired_space,
```
Comment (mdhaber) [inline scipy/_lib/_array_api.py:223]: I've heard preferences for plain `assert` these days. I'm not sure if that is still true when you want the assertion to have a custom message, but you *can* produce a custom message like `assert 0, "message"`.

File: scipy/_lib/_array_api.py [hunk 5]
```diff
@@ -191,33 +192,81 @@ def is_torch(xp):
...
 
 
-def assert_equal(actual, desired, err_msg='', xp=None):
+def _strict_check(actual, desired, check_xp=True, check_dtype=True, check_shape=True):
+    try:
+        actual.dtype
+        desired.dtype
+    except AttributeError:
+        raise ValueError("Inputs should be arrays with a dtype attribute, "
+                         "python scalars and arrays are not accepted.")
```
Comment (mdhaber) [inline scipy/_lib/_array_api.py:201]: This is even more strict than I anticipated - perhaps uncomfortably strict even for me... but I think I can get over it. I'll reply again after thinking  bit. I'm guessing that I'll suggest for this check to be part of `check_xp`. If the user disables `check_xp`, at least `desired` can be automatically be made a NumPy type like `desired = np.asarray(desired)[()]`.

File: scipy/_lib/_array_api.py [hunk 6]
```diff
@@ -191,33 +192,81 @@ def is_torch(xp):
...
+
+
+def _check_scalar(actual, desired):
+    if actual.shape == ():
+        if isinstance(actual, np.ndarray):
+            assert_(isinstance(desired, np.ndarray),
+                    "Desired a numpy scalar, but 0-D array given.")
+        else:
+            assert_(not isinstance(desired, np.ndarray),
+                    "Desired a 0-D array, but numpy scalar given.")
```
Comment (mdhaber) [inline scipy/_lib/_array_api.py:237]:
```suggestion
    if np.isscalar(desired):
        assert_(np.isscalar(actual),
                "Desired a scalar, but 0-D array given.")
    else:
        assert_(not np.isscalar(actual),
                "Desired a 0-D array, but scalar given.")
```

File: scipy/_lib/_array_api.py (hunk 4 above)
Comment (mdhaber) [reply scipy/_lib/_array_api.py:223]: Never mind, I see that you can't do plain `assert` with implicit line break.

File: scipy/_lib/_array_api.py (hunk 3 above)
Comment (lucascolley) [reply scipy/_lib/_array_api.py:215]: I don't think that it is currently used with tuple `actual` anywhere, I'm not sure about the future.

> actual = actual if isinstance(actual, tuple) else (actual,)

Sounds good.

Comment (lucascolley) [reply scipy/_lib/_array_api.py:215]: > the assertions are written differently, and one has a message while the other doesn't. (Or was this intentional?)

Oops, that is a bug, the custom message should apply for both.

File: scipy/_lib/_array_api.py (hunk 5 above)
Comment (mdhaber) [reply scipy/_lib/_array_api.py:201]: Since the purpose of `check_xp` is to allow the developer to disable strict type checking, I would assume that with `check_xp=False`, a Python float would be OK. 

However, all the checks after this rely on `dtype` and `shape` attributes, so either they would fail, or the Python float would need to become a float (or 0d array) of the appropriate library.

Maybe if `check_xp=False`, then:
```python3
xp = array_namespace(actual)
desired = xp.asarray(desired)[()]
```
?

File: scipy/_lib/_array_api.py [hunk 7]
```diff
@@ -195,33 +196,78 @@ def is_torch(xp):
...
+            arr_space = array_namespace(arr)
+            assert arr_space == desired_space
+            if is_numpy(arr_space):
+                _check_scalar(actual, desired)
+    else:
+        actual_space = array_namespace(actual)
+        assert_(actual_space == desired_space,
+                "Namespaces do not match: "
+                f"{actual_space.__name__} and {desired_space.__name__}")
+        if is_numpy(actual_space):
```
Comment (mdhaber) [inline scipy/_lib/_array_api.py:230]: It looks like CuPy makes a distinction between scalars and 0d array, too. However, it could be argued that testing for scalars vs arrays is not as important for CuPy as it is for NumPy because it is much harder to get a scalar out of CuPy:
```python3
import cupyx
cupyx.scipy.special.ndtr(0)[()]
```
That is, in CuPy, functions return a 0d array even when given a scalar, and using `[()]` doesn't extract a scalar, so it's much harder to get inconsistencies accidentally.

File: scipy/_lib/_array_api.py (hunk 3 above)
Comment (lucascolley) [reply scipy/_lib/_array_api.py]: I've addressed this in https://github.com/scipy/scipy/pull/19186/commits/2df7541e02f45415d8ecdc99b0e8f64a5af3043d. cc @tylerjereddy as original author.

File: scipy/_lib/_array_api.py (hunk 7 above)
Comment (lucascolley) [reply scipy/_lib/_array_api.py]: It does only apply to NumPy. I think that's probably fine, but if you want to check for other libraries as well then we could.

File: scipy/_lib/_array_api.py (hunk 5 above)
Comment (lucascolley) [reply scipy/_lib/_array_api.py:205]: - will `xp.asarray(desired)[()]` work for `torch` too?
- will it just be `xp.asarray(desired)` if we have a list instead of a scalar? If so, what function is used to check what we have?
- What about when `actual` is a 0-D array?

Comment (mdhaber) [reply scipy/_lib/_array_api.py:205]: > will `xp.asarray(desired)[()]` work for torch too?

It won't fail. This will always convert `desired` to an array except if `desired` is 0d and `xp` is NumPy. In that case, it will convert `desired` to a NumPy scalar.

> will it just be `xp.asarray(desired)` if we have a list instead of a scalar? If so, what function is used to check what we have?

`[()]` doesn't extract items out of an array with more than 0 dimensions, so I don't think you need to have a special case for lists vs scalars.

> What about when `actual` is a 0-D array?

I don't understand. `array_namespace(actual)` will still work even if `actual` is a 0d array, right? 

---

Perhaps all this is getting at what happens if `desired` is a 0D array? That is the only case I see in which this suggestion might be problematic because (for NumPy only) it would convert the 0d array to a scalar. If you don't want that, you could check whether `desired` is a NumPy array, and only then do the `[()]`. 

I neglected to include that step on purpose because IMO, SciPy functions should follow whatever NumPy's convention is, and vanilla NumPy's convention is to essentially *never* return a 0d array.
```python3
np.mean([1, 2, 3])  # scalar, not 0d array
np.asarray(0)*2  # scalar, not 0d array
np.sin(np.asarray(0))  # scalar, not 0d array
```
Strange as that may be, I think we should follow suit. So IF the developer were to pass a 0d NumPy array for `desired`, that would almost certainly be a mistake (since `actual` should never be a NumPy 0d array), and we are doing them the favor of converting the 0d array to a scalar. Then, this wil check to see whether `actual` is a scalar.

If NumPy's convention changes to eliminate scalars and start returning 0d arrays, then `xp.asarray(desired)[()]` will begin to return a 0d array, and that's fine, too. `desired` will become a 0d array, and we'll be checking whether `actual` is a 0d array. 

In summary, this will track the convention of `xp`: if `[()]` returns a 0d array, then we would want `actual` to be a 0d array instead of a scalar, and if `[()]` returns a scalar, we would want `actual` to be a scalar instead of a 0d array.

Comment (mdhaber) [reply scipy/_lib/_array_api.py:205]: If it helps, I guess `[()]` is not needed because this only happens *after* the `check_xp` block, and `[()]` would change neither the shape nor dtype.

Comment (lucascolley) [reply scipy/_lib/_array_api.py:205]: > assert_equal(0, np.asarray(0)[()])  # fails - good!

I'm unable to reproduce this failure from your example above... are you sure?

I don't know how to go about making this fail either.

Comment (lucascolley) [reply scipy/_lib/_array_api.py]: After https://github.com/scipy/scipy/pull/19186/commits/b99f0e716450eaae5134ceea8610d0d6e50a30b7, we now have improved behaviour with `check_xp=False`, e.g. `xp_assert_equal(np.asarray([1, 2]), torch.asarray([1, 2]), check_xp=False)` passes.

I have included `actual = xp.asarray(actual)` to enable dtype checks when `actual` is a built-in type.

However, I am struggling with built-in types for `check_xp=True`. `xp_assert_equal(np.asarray(0)[()], 0)`, `xp_assert_equal(0, np.asarray(0)[()])` and `xp_assert_equal(np.asarray([1, 2]), [1, 2])` all pass.

Also, the error message for `xp_assert_equal(1, [1, 2])` is `Desired an ndarray, but scalar given.`, which is not quite correct. We desire a list. This passes the rest of the function since `array_namespace` returns `array_api_compat.numpy` for all array-likes now.

Feel free to commit any changes that you need to get this working Matt!

Comment (mdhaber) [reply scipy/_lib/_array_api.py]: > I'm unable to reproduce this failure from your example above... are you sure?

No I can't reproduce that. Not sure what I was doing before...

File: scipy/_lib/_array_api.py (hunk 7 above)
Comment (rgommers) [reply scipy/_lib/_array_api.py]: > It looks like CuPy makes a distinction between scalars and 0d array, too.

There are no scalars in CuPy, Pytorch or any other relevant library. They all use 0-D arrays and that works just fine. We'd remove scalars from NumPy if we could, but that ship has sailed.

**Comment by lucascolley** [comment]
This looks almost done now, cc @tylerjereddy @rgommers if you'd like to have a look.

**Comment by lucascolley** [comment]
@mdhaber it looks like `_lazywhere` may need to be updated to satisfy the expected 0D-array / scalar behaviour?

**Review by mdhaber — APPROVED** [review]
I'm happy with this, but it could use another review since I've been pretty heavily involved. Some notable points are:
- function signatures
- changes to `_lib._array_api.cov` (unrelated to the rest of the PR, but I happened to notice some apparently dead code while working on this.)
- When `check_shape=True`, `_check_scalar` raises an `AssertionError` if a function returns a NumPy 0d array instead of a NumPy scalar. (This is specific to NumPy; 0d arrays of any other type are allowed/enforced.) This is to encourage consistency with NumPy, which returns a scalar instead of a 0d array at almost every opportunity. Noteable examples include:
  - `np.mean([1, 2, 3])`
  - `np.asarray(1)*2`
  -  `np.asarray(1) + np.asarray(1)`
  - `np.sin(np.asarray(1))`

I know that not all of SciPy follows this convention, but most code I have worked with in `stats`, `optimize`, `special`, and `integrate` does, so I think we should include this assertion by default to work toward better self-consistency and consistency with NumPy in new, Array API compatible features.

File: scipy/_lib/_array_api.py [hunk 8]
```diff
@@ -12,6 +12,7 @@
 import warnings
 
 import numpy as np
+from numpy.testing import assert_
```
Comment (tylerjereddy) [inline scipy/_lib/_array_api.py:15]: We have a preference for not using this `assert_` and instead just using the plain `pytest` `assert` at this point, though I forget where we documented that.

In any case, please don't make that change at this point, I'm just doing a hopefully-final pass...

Comment (lucascolley) [reply scipy/_lib/_array_api.py:15]: https://github.com/scipy/scipy/pull/19186#discussion_r1320077302 makes sense, but we have a line break, so I think we want `assert_`? Or `msg=...` on the line beforehand might be better.

Comment (tylerjereddy) [reply scipy/_lib/_array_api.py:15]: I'd just split the message assignment yeah, but not worth it at this point of code review IMO.

Comment (mdhaber) [reply scipy/_lib/_array_api.py:15]: Yup, I noted this earlier, but let it pass to avoid the explicit line breaks.

**Comment by tylerjereddy** [comment]
I'll merge this shortly, probably squash-merge given 20 commits and some 1-liners in there, etc.

Things I checked/other thoughts:

1. Full test suite with no array API stuff on.
2. Full test suite with array API + cuda. (there's a known `torch` failure already fixed upstream)
3. Full test suite with array API on host.
4. There is a comment above that mentions a few cases we may to make some adjustment for, including this one that doesn't pass (raise an exception) for current non-conformant NumPy (but does for all other backends)--perhaps we should open a separate issue for those few cases if one isn't open already (I don't think I know the answer to what we want for all those cases though...):

```diff
--- a/scipy/_lib/tests/test_array_api.py
+++ b/scipy/_lib/tests/test_array_api.py
@@ -106,3 +106,9 @@ class TestArrayAPI:
             with pytest.raises(AssertionError, match="Types do not match."):
                 xp_assert_equal(xp.asarray(0.), xp.float64(0))
             xp_assert_equal(xp.float64(0), xp.asarray(0.))
+
+
+    @array_api_compatible
+    def test_coercion_stringency(self, xp):
+        with pytest.raises(AssertionError):
+            xp_assert_equal(xp.asarray([1, 2]), [1, 2])
```

5. So far, upstream (NumPy) doesn't seem to object too much to adding some similar stringency options to their current assertions like Matt's PR here: https://github.com/numpy/numpy/pull/24680 (and they already did it elsewhere as discussed in previous comments)
6. `xp_assert_close` isn't tested/used yet--I did note when substituting it in a few `assert_allclose()` places that it doesn't support `float` input with default arguments--maybe that's an "ok" usage for the conventional NumPy assertion machinery I suppose. (`AttributeError: 'float' object has no attribute 'dtype'`). Tests should be easy enough to write like this if we want:

```diff
--- a/scipy/_lib/tests/test_array_api.py
+++ b/scipy/_lib/tests/test_array_api.py
@@ -3,7 +3,8 @@ import pytest
 
 from scipy.conftest import array_api_compatible
 from scipy._lib._array_api import (
-    _GLOBAL_CONFIG, array_namespace, as_xparray, copy, xp_assert_equal, is_numpy
+    _GLOBAL_CONFIG, array_namespace, as_xparray, copy, xp_assert_equal, is_numpy,
+    xp_assert_close,
 )
 import scipy._lib.array_api_compat.array_api_compat.numpy as np_compat
 
@@ -106,3 +107,9 @@ class TestArrayAPI:
             with pytest.raises(AssertionError, match="Types do not match."):
                 xp_assert_equal(xp.asarray(0.), xp.float64(0))
             xp_assert_equal(xp.float64(0), xp.asarray(0.))
+
+    @array_api_compatible
+    def test_assert_close(self, xp):
+        with pytest.raises(AssertionError):
+            xp_assert_close(xp.asarray([0.0001, 0.0001]), xp.asarray([0.00009, 0.00009]))
+        xp_assert_close(xp.asarray([0.0001, 0.0001]), xp.asarray([0.00009, 0.00009]), atol=1e-05)

```

I imagine a few rough spots will get ironed out when these assertions start getting used more heavily.

**Comment by mdhaber** [comment]
> it doesn't support float input with default arguments

This was intentional, given the discussion in https://github.com/scipy/scipy/pull/19157#discussion_r1312983982.

**Comment by lucascolley** [comment]
Thanks for the thorough review Tyler, and all of the help on this one Matt!

**Comment by tylerjereddy** [comment]
> However, if check_namespace is turned off, it should accept plain floats and do the conversion automatically.

Maybe, although if they are *both* floats you're still hosed it seems:

```diff
--- a/scipy/spatial/tests/test_hausdorff.py
+++ b/scipy/spatial/tests/test_hausdorff.py
@@ -6,6 +6,7 @@ import pytest
 from scipy.spatial.distance import directed_hausdorff
 from scipy.spatial import distance
 from scipy._lib._util import check_random_state
+from scipy._lib._array_api import xp_assert_close
 
 
 class TestHausdorff:
@@ -146,7 +147,7 @@ class TestHausdorff:
         # verify fix for gh-11332
         actual = directed_hausdorff(u=A, v=B, seed=seed)
         # check distance
-        assert_allclose(actual[0], expected[0])
+        xp_assert_close(actual[0], expected[0], check_namespace=False)
```

Since that functionality probably isn't even planned for array API support anytime soon it doesn't matter much. On the flip side, I was perhaps wondering if longer-term we'd want to consider completely switching over to that set of assertions everywhere and/or having the error message explain what to do in that case.

**Comment by mdhaber** [comment]
> although if they are both floats you're still hosed it seems

Sure. That's just a consequence of the way https://github.com/scipy/scipy/pull/19157#discussion_r1312983982 went. 

> On the flip side, I was perhaps wondering if longer-term we'd want to consider completely switching over

Sounds great to me.
